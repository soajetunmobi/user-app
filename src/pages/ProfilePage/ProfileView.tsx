import React, { FunctionComponent } from 'react'
import { Button, createStyles, Grid } from '@material-ui/core'
import { makeStyles } from '@material-ui/core/styles'
import { LoadingIndicator } from 'components/shared/LoadingIndicator'

const useStyles = makeStyles(() =>
  createStyles({
    root: {
      maxWidth: '450px',
      display: 'block',
      margin: '0 auto'
    },
    textField: {
      '& > *': {
        width: '100%'
      }
    },
    submitButton: {
      marginTop: '24px'
    },
    title: { textAlign: 'center' },
    successMessage: { color: 'green' },
    errorMessage: { color: 'red' }
  })
)

interface ProfileViewProps {
  user: User
  cancelView: () => void
  editContent: () => void
}

export const ProfileView: FunctionComponent<ProfileViewProps> = ({
  user,
  cancelView,
  editContent
}) => {
  const classes = useStyles()

  const renderAddress = () => {
    const { address } = user
    return (
      <div>
        <div>Street: {address.street}</div>
        <div>Town: {address.town}</div>
        <div>County: {address.county}</div>
        <div>Postcode: {address.postcode}</div>
      </div>
    )
  }
  const renderPreferences = () => {
    const {
      preferences: { contact }
    } = user
    return (
      <div>
        {contact.map((choice: preference) => {
          return <span key={choice}>{choice}</span>
        })}
      </div>
    )
  }

  if (!user) {
    return <LoadingIndicator />
  }

  return (
    <div className={classes.root} data-testid='profile-view'>
      <h1 className={classes.title}>Profile - View</h1>
      <Grid container justify='space-around' direction='row'>
        <Grid item lg={10} md={10} sm={10} xs={10} className={classes.textField}>
          <span>Firstname:</span>
          <span data-testid='profile-firstname'>{user.first_name}</span>
        </Grid>
        <Grid item lg={10} md={10} sm={10} xs={10} className={classes.textField}>
          <span>Other names:</span>
          <span data-testid='profile-othernames'>{user.other_names}</span>
        </Grid>
        <Grid item lg={10} md={10} sm={10} xs={10} className={classes.textField}>
          <span>Address:</span>
          {renderAddress()}
        </Grid>
        <Grid item lg={10} md={10} sm={10} xs={10} className={classes.textField}>
          <span>Mobile:</span>
          <span>{user.mobile}</span>
        </Grid>
        <Grid item lg={10} md={10} sm={10} xs={10} className={classes.textField}>
          <span>Email:</span>
          <span>{user.mobile}</span>
        </Grid>
        <Grid item lg={10} md={10} sm={10} xs={10} className={classes.textField}>
          <span>Company:</span>
          <span>{user.company}</span>
        </Grid>
        <Grid item lg={10} md={10} sm={10} xs={10} className={classes.textField}>
          <span>Preferences:</span>
          {renderPreferences()}
        </Grid>
        <Grid item lg={10} md={10} sm={10} xs={10} className={classes.submitButton}>
          <Button
            type='submit'
            variant='outlined'
            color='secondary'
            data-testid='profile-cancel-button'
            onClick={cancelView}
          >
            Cancel
          </Button>
          <Button
            type='submit'
            variant='contained'
            color='secondary'
            data-testid='profile-edit-button'
            onClick={editContent}
          >
            Edit
          </Button>
        </Grid>
      </Grid>
    </div>
  )
}
