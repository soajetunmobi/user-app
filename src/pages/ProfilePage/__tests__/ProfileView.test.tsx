import { MOCKED_USER } from 'testUtils/mocks/user'
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React from 'react'
import { ProfileView } from '../ProfileView'

describe('ProfileView', () => {
  it('profile view is rendered', () => {
    const { asFragment } = render(
      <ProfileView user={MOCKED_USER} cancelView={jest.fn} editContent={jest.fn} />
    )
    expect(asFragment).toMatchSnapshot()
  })

  describe('Profile view functionality', () => {
    const editButtonMock = jest.fn()
    const cancelButtonMock = jest.fn()

    it('should be able to edit mode', () => {
      render(
        <ProfileView
          user={MOCKED_USER}
          editContent={editButtonMock}
          cancelView={cancelButtonMock}
        />
      )
      userEvent.click(screen.getByTestId('profile-edit-button'))
      expect(editButtonMock).toHaveBeenCalled()
      expect(editButtonMock).toHaveBeenCalledTimes(1)
      expect(cancelButtonMock).not.toHaveBeenCalled()
    })

    it('should be able to cancel viewing', () => {
      render(
        <ProfileView
          user={MOCKED_USER}
          editContent={editButtonMock}
          cancelView={cancelButtonMock}
        />
      )
      userEvent.click(screen.getByTestId('profile-cancel-button'))
      expect(cancelButtonMock).toHaveBeenCalled()
      expect(cancelButtonMock).toHaveBeenCalledTimes(1)
      expect(editButtonMock).not.toHaveBeenCalled()
    })

    it('should see profile information on screen', () => {
      const { queryByTestId } = render(
        <ProfileView
          user={MOCKED_USER}
          editContent={editButtonMock}
          cancelView={cancelButtonMock}
        />
      )
      expect(queryByTestId('profile-firstname')).toContainHTML(MOCKED_USER.first_name)
      expect(queryByTestId('profile-othernames')).toContainHTML(MOCKED_USER.other_names)
    })
  })
})
