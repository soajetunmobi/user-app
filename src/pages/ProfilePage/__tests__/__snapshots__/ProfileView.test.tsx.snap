// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileView profile view is rendered 1`] = `[Function]`;
